src/StateFuture.ts

```diff
@@ -0,0 +1,303 @@
+export const modify: <S, E = never>(f: (s: S) => S) => StateFuture<S, E, void> = T.modify
+
+/**
+ * @since 0.6.5
+ */
+export const gets: <S, E = never, A = never>(f: (s: S) => A) => StateFuture<S, E, A> = T.gets
+
+/**
+ * @since 0.6.5
+ */
+export function fromEitherK<E, A extends Array<unknown>, B>(
+  f: (...a: A) => E.Either<E, B>
+): <S>(...a: A) => StateFuture<S, E, B> {
+  return (...a) => fromEither(f(...a))
+}
+
+/**
+ * @since 0.6.5
+ */
+export function chainEitherK<E, A, B>(
+  f: (a: A) => E.Either<E, B>
+): <S>(ma: StateFuture<S, E, A>) => StateFuture<S, E, B> {
+  return chain<any, E, A, B>(fromEitherK(f))
+}
+
+/**
+ * @since 0.6.5
+ */
+export function fromFutureK<E, A extends Array<unknown>, B>(
+  f: (...a: A) => F.Future<E, B>
+): <S>(...a: A) => StateFuture<S, E, B> {
+  return (...a) => fromFuture(f(...a))
+}
+
+/**
+ * @since 0.6.5
+ */
+export function chainFutureK<E, A, B>(
+  f: (a: A) => F.Future<E, B>
+): <S>(ma: StateFuture<S, E, A>) => StateFuture<S, E, B> {
+  return chain<any, E, A, B>(fromFutureK(f))
+}
+
+/**
+ * @since 0.6.5
+ */
+export const stateFuture: Monad3<URI> &
+  MonadThrow3<URI> &
+  Bifunctor3<URI> &
+  Alt3<URI> &
+  ChainRec3<URI> &
+  MonadThrow3<URI> &
+  MonadTask3<URI> = {
+  URI,
+  map: T.map,
+  of: right,
+  ap: T.ap,
+  chain: T.chain,
+  throwError: left,
+  alt: (fx, f) => c =>
+    pipe(
+      fx(c),
+      F.alt(() => f()(c))
+    ),
+  bimap: (fea, f, g) => c =>
+    pipe(
+      fea(c),
+      F.bimap(f, ([a, c]) => [g(a), c])
+    ),
+  chainRec: <S, E, A, B>(a: A, f: (a: A) => StateFuture<S, E, E.Either<A, B>>): StateFuture<S, E, B> =>
+    (function recur(a: A): StateFuture<S, E, B> {
+      return stateFuture.chain(f(a), E.fold(recur, right))
+    })(a),
+  mapLeft: (fea, f) => c => pipe(fea(c), F.mapLeft(f)),
+  fromIO: rightIO,
+  fromTask: rightTask
+}
+
+/**
+ * Like `stateFuture` but `ap` is sequential
+ * @since 0.6.5
+ */
+export const stateFutureSeq: typeof stateFuture = {
+  ...stateFuture,
+  ap: (mab, ma) => T.chain(mab, f => T.map(ma, f))
+}
+
+const {
+  alt,
+  ap,
+  apFirst,
+  apSecond,
+  bimap,
+  chain,
+  chainFirst,
+  flatten,
+  map,
+  mapLeft,
+  filterOrElse,
+  fromEither,
+  fromOption,
+  fromPredicate
+} = pipeable(stateFuture)
+
+const chainRec = stateFuture.chainRec
+
+export {
+  /**
+   * @since 0.6.5
+   */
+  alt,
+  /**
+   * @since 0.6.5
+   */
+  ap,
+  /**
+   * @since 0.6.5
+   */
+  apFirst,
+  /**
+   * @since 0.6.5
+   */
+  apSecond,
+  /**
+   * @since 0.6.5
+   */
+  bimap,
+  /**
+   * @since 0.6.5
+   */
+  chain,
+  /**
+   * @since 0.6.5
+   */
+  chainFirst,
+  /**
+   * @since 0.6.5
+   */
+  flatten,
+  /**
+   * @since 0.6.5
+   */
+  map,
+  /**
+   * @since 0.6.5
+   */
+  mapLeft,
+  /**
+   * @since 0.6.5
+   */
+  filterOrElse,
+  /**
+   * @since 0.6.5
+   */
+  fromEither,
+  /**
+   * @since 0.6.5
+   */
+  fromOption,
+  /**
+   * @since 0.6.5
+   */
+  fromPredicate,
+  /**
+   * @since 0.6.5
+   */
+  chainRec
```

Comment:
not sure why you are exporting `chainRec` here

Reply:
Right, this is not really necessary.

By the way, this addition is actually related to a module that I'm currently working on.

I used `hyper-ts/Middleware` as starting example , wish I had seen `StateTaskEither` a bit earlier.

`chainRec` requirement can be seen [here](https://github.com/tetsuo/hippo/blob/1.x/src/Parser.ts), but `future.chainRec` also works fine of course.

(`hippo` is like `hyper-ts` for `net` sockets- less type-safe though)

I use `chainRec` with a `parser-ts/Stream` to parse the contents of node-js stream.

Simply told, on each step, I read some data from the stream's internal buffer and keep doing that until parsing completes (or raises an error), and it works really fine actually for simple things 👍 That would have been much efficient if `parser-ts` gave some length information so I could wait for buffer to fill up, and, then, run the parser.  But `hippo` module [will also provide](https://github.com/tetsuo/hippo/blob/1.x/src/Decoder.ts) some length-prefixing essentials so I hope that it won't be necessary to put that logic inside `parser-ts`.